### PR TITLE
Fixes bug in JoinNode populate when using `addFromModel` to populate DataModel

### PR DESCRIPTION
**Pull request by bhargav**
When copying instances using the `DataModel.addFromModel` method, JoinNodes were copied and also populated recursively. This lead to a large number of unneeded Join Nodes being created during DataModel instance population.

Fixing this issue and handling it similar to the `populateEdge` parameter in this flow.

Addresses #449

### CODE REVIEW

File: saul-examples/src/main/scala/edu/illinois/cs/cogcomp/saulexamples/nlp/SemanticRoleLabeling/PopulateSRLDataModel.scala [hunk 1]
```diff
@@ -111,7 +110,6 @@ object PopulateSRLDataModel extends Logging {
         logger.debug("gold relations for this train:" + gr.relations().size)
         if (!useGoldArgBoundaries) {
           val XuPalmerCandidateArgsTraining = gr.predicates.getTrainingInstances.flatMap(x => xuPalmerCandidate(x, (gr.sentences(x.getTextAnnotation) ~> gr.sentencesToStringTree).head))
-          gr.sentencesToRelations.addSensor(textAnnotationToRelationMatch _)
```
Comment (kordjamshidi) [inline saul-examples/src/main/scala/edu/illinois/cs/cogcomp/saulexamples/nlp/SemanticRoleLabeling/PopulateSRLDataModel.scala]: don't  we need this sensor? don't we get duplicate relations?

**Comment by kordjamshidi** [comment]
Thank you @bhargav . This was a very good catch.

File: saul-examples/src/main/scala/edu/illinois/cs/cogcomp/saulexamples/nlp/SemanticRoleLabeling/PopulateSRLDataModel.scala (hunk 1 above)
Comment (bhargav) [inline saul-examples/src/main/scala/edu/illinois/cs/cogcomp/saulexamples/nlp/SemanticRoleLabeling/PopulateSRLDataModel.scala]: We already have this matching sensor in the DataModel. https://github.com/IllinoisCogComp/saul/blob/master/saul-examples/src/main/scala/edu/illinois/cs/cogcomp/saulexamples/nlp/SemanticRoleLabeling/SRLMultiGraphDataModel.scala#L51

Comment (kordjamshidi) [inline saul-examples/src/main/scala/edu/illinois/cs/cogcomp/saulexamples/nlp/SemanticRoleLabeling/PopulateSRLDataModel.scala]: I see, you are right. Thanks.

File: saul-core/src/main/scala/edu/illinois/cs/cogcomp/saul/datamodel/node/Node.scala [hunk 2]
```diff
@@ -126,19 +132,30 @@ class Node[T <: AnyRef](val keyFunc: T => Any = (x: T) => x, val tag: ClassTag[T
...
 
-  def populateFrom(n: Node[_]): Unit = {
-    populate(n.getTrainingInstances.map(_.asInstanceOf[T]), train = true, populateEdge = false)
-    populate(n.getTestingInstances.map(_.asInstanceOf[T]), train = false, populateEdge = false)
+  /** Populate current node instances using instances from a different node on the same type.
+    * Usage Node: Expected usage is to copy nodes from one DataModel to another.
+    *
+    * @param n Node to populate instances from.
+    */
+  def populateFrom(n: Node[_])(implicit tag: ClassTag[instanceType]): Unit = {
```
Comment (danyaljj) [inline saul-core/src/main/scala/edu/illinois/cs/cogcomp/saul/datamodel/node/Node.scala]: Possibly for future: we have to clarify its use case in the documentation.

File: saul-examples/src/main/scala/edu/illinois/cs/cogcomp/saulexamples/nlp/SemanticRoleLabeling/PopulateSRLDataModel.scala [hunk 3]
```diff
@@ -54,7 +54,7 @@ object PopulateSRLDataModel extends Logging {
       case ViewNames.PARSE_GOLD => new ClauseViewGenerator(parseViewName, "CLAUSES_GOLD")
       case ViewNames.PARSE_STANFORD => ClauseViewGenerator.STANFORD
     }
-    def addViewAndFilter(taAll: List[TextAnnotation]): List[TextAnnotation] = {
+    def addViewAndFilter(taAll: Iterable[TextAnnotation]): Iterable[TextAnnotation] = {
```
Comment (danyaljj) [inline saul-examples/src/main/scala/edu/illinois/cs/cogcomp/saulexamples/nlp/SemanticRoleLabeling/PopulateSRLDataModel.scala]: I like this change. You probably want to drop `toList` from [here](https://github.com/IllinoisCogComp/saul/blob/765f480d5e9d1896e319775ba0479e31a98081fc/saul-examples/src/main/scala/edu/illinois/cs/cogcomp/saulexamples/nlp/SemanticRoleLabeling/PopulateSRLDataModel.scala#L96) too.

File: saul-examples/src/main/scala/edu/illinois/cs/cogcomp/saulexamples/nlp/SemanticRoleLabeling/SRLMultiGraphDataModel.scala [hunk 4]
```diff
@@ -14,6 +14,7 @@ import edu.illinois.cs.cogcomp.nlp.corpusreaders.AbstractSRLAnnotationReader
 import edu.illinois.cs.cogcomp.saul.datamodel.DataModel
 import edu.illinois.cs.cogcomp.saul.datamodel.property.PairwiseConjunction
 import edu.illinois.cs.cogcomp.saulexamples.data.SRLFrameManager
+import edu.illinois.cs.cogcomp.saulexamples.nlp.CommonSensors.textAnnotationToTokens
```
Comment (danyaljj) [inline saul-examples/src/main/scala/edu/illinois/cs/cogcomp/saulexamples/nlp/SemanticRoleLabeling/SRLMultiGraphDataModel.scala]: Not very related to this PR. Could you capitalize this class?

**Comment by bhargav** [comment]
Update:

- Remove unused `joinNode` in the SRL Example
- Add a basic unit test for JoinNode population

File: saul-core/src/main/scala/edu/illinois/cs/cogcomp/saul/datamodel/node/JoinNode.scala [hunk 5]
```diff
@@ -34,12 +34,12 @@ class JoinNode[A <: AnyRef, B <: AnyRef](val na: Node[A], val nb: Node[B], match
     }
   }
 
-  override def addInstance(t: (A, B), train: Boolean, populateEdge: Boolean = true): Unit = {
+  override def addInstance(t: (A, B), train: Boolean, populateEdge: Boolean = true, populateJoinNodes: Boolean = true): Unit = {
```
Comment (bhargav) [inline saul-core/src/main/scala/edu/illinois/cs/cogcomp/saul/datamodel/node/JoinNode.scala]: Will handle the documentation related comments as a separate PR.

File: saul-core/src/main/scala/edu/illinois/cs/cogcomp/saul/datamodel/node/Node.scala [hunk 6]
```diff
@@ -126,19 +132,30 @@ class Node[T <: AnyRef](val keyFunc: T => Any = (x: T) => x, val tag: ClassTag[T
         incoming.foreach(_.populateUsingTo(instance, train))
       }
 
-      // TODO: Populating join nodes takes significant amount of time on large graphs. Investigate.
```
Comment (bhargav) [inline saul-core/src/main/scala/edu/illinois/cs/cogcomp/saul/datamodel/node/Node.scala]: Added the TODO back.

**Pull request merged**